cpp/src/arrow/util/async_generator.h

```diff
@@ -1058,18 +1109,73 @@ class MergedGenerator {
       return source();
     }
 
+    void SignalErrorUnlocked() {
+      broken = true;
+      // Empty any results that have arrived but not asked for.
+      while (!delivered_jobs.empty()) {
+        delivered_jobs.pop_front();
+      }
+    }
+
+    void Purge() {
+      while (!waiting_jobs.empty()) {
+        waiting_jobs.front()->MarkFinished(IterationEnd<T>());
+        waiting_jobs.pop_front();
+      }
+    }
+
+    void MarkFinished() {
```

Comment:
it seems sometimes we MarkFinished directly, and other times we use this function - maybe this should be renamed `MarkFinishedAndPurge` or something to make it clear why they're different?

Reply:
Done.